clang/lib/Sema/SemaDecl.cpp

```diff
@@ -7678,13 +7678,34 @@ NamedDecl *Sema::ActOnVariableDeclarator(
       NewVD->setTSCSpec(TSCS);
   }
 
-  // Global variables with types decorated with device_global attribute must be
-  // static if they are declared in SYCL device code.
   if (getLangOpts().SYCLIsDevice) {
-    if (SCSpec != DeclSpec::SCS_static && !NewVD->hasGlobalStorage() &&
-        isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(
-            NewVD->getType()))
-      Diag(D.getIdentifierLoc(), diag::err_sycl_device_global_incorrect_scope);
+    // device_global array is not allowed.
+    if (NewVD->getType()->isArrayType()) {
+      if (const auto *AT = dyn_cast<ArrayType>(NewVD->getType())) {
```

Comment:
Can you elaborate on why its not safe @Fznamznon ? Actually I just noticed the `isArrayType( )` and would have suggested removing L6784 prior to your comment.

Reply:
I spoke to Aaron briefly and I believe it is the fact that this is an array type which makes this tricky. Other types can usually be `dyn_cast` without issue I think (Please correct me if I am mistaken here @Fznamznon ). Aaron suggested using `ASTContext::getAsArrayType()` to check for array type if you plan to reuse type later. 

Also, I've been told (when my code was reviewed in the past) that it is generally considered bad practice in community to do a `isXYZType( )` followed by a `getXYZType()`  or cast as done here, since the former is redundant in most cases.

Reply:
>  Can you elaborate on why its not safe @Fznamznon ?

I think we've already had some troubles with `dyn_cast`-ing types a while ago. See https://github.com/intel/llvm/pull/2072#discussion_r452201923 .

> Actually I just noticed the isArrayType( ) and would have suggested removing L6784 prior to your comment.

This is what I was trying to suggest: remove the line with dyn_cast (L7684) since there is `isArrayType()` already, use `NewVD->getType()->getPointeeOrArrayElementType()` later in the code.

Reply:
Thanks for the link and catching this! I had forgotten about this! Sorry for giving you incorrect information earlier @schittir. Can you change the code here as suggested?